src/main/java/baritone/cache/FasterWorldScanner.java

```diff
@@ -0,0 +1,249 @@
+            if (((IBlockStateContainer) sectionContainer).getStorage() == null) {
+                continue;
+            }
+
+            boolean[] isInFilter = getIncludedFilterIndices(lookup, ((IBlockStateContainer) sectionContainer).getPalette());
+            if (isInFilter.length == 0) {
+                continue;
+            }
+            consumer.accept(section, isInFilter);
+        }
+    }
+
+    private boolean[] getIncludedFilterIndices(BlockOptionalMetaLookup lookup, IBlockStatePalette palette) {
+        boolean commonBlockFound = false;
+        ObjectIntIdentityMap<IBlockState> paletteMap = getPalette(palette);
+        int size = paletteMap.size();
+
+        boolean[] isInFilter = new boolean[size];
+
+        for (int i = 0; i < size; i++) {
+            IBlockState state = paletteMap.getByValue(i);
+            if (lookup.has(state)) {
+                isInFilter[i] = true;
+                commonBlockFound = true;
+            } else {
+                isInFilter[i] = false;
+            }
+        }
+
+        if (!commonBlockFound) {
+            return new boolean[0];
+        }
+        return isInFilter;
+    }
+
+    private static void collectBlockLocations(BitArray array, boolean[] isInFilter, IntConsumer action) {
+        long[] longArray = array.getBackingLongArray();
+        int arraySize = array.size();
+        int bitsPerEntry = ((IBitArray) array).getBitsPerEntry();
+        long maxEntryValue = ((IBitArray) array).getMaxEntryValue();
+
+        for (int idx = 0, kl = bitsPerEntry - 1; idx < arraySize; idx++, kl += bitsPerEntry) {
+            final int i = idx * bitsPerEntry;
+            final int j = i >> 6;
+            final int l = i & 63;
+            final int k = kl >> 6;
+            final long jl = longArray[j] >>> l;
+
+            if (j == k) {
+                if (isInFilter[(int) (jl & maxEntryValue)]) {
+                    action.accept(idx);
+                }
+            } else {
+                if (isInFilter[(int) ((jl | longArray[k] << (64 - l)) & maxEntryValue)]) {
+                    action.accept(idx);
+                }
+            }
+        }
+    }
+
+    /**
+     * cheats to get the actual map of id -> blockstate from the various palette implementations
+     */
+    private static ObjectIntIdentityMap<IBlockState> getPalette(IBlockStatePalette palette) {
+        if (palette instanceof BlockStatePaletteRegistry) {
+            return Block.BLOCK_STATE_IDS;
+        } else {
+            PacketBuffer buf = new PacketBuffer(Unpooled.buffer());
+            palette.write(buf);
+            int size = buf.readVarInt();
+            ObjectIntIdentityMap<IBlockState> states = new ObjectIntIdentityMap<>();
+            for (int i = 0; i < size; i++) {
+                IBlockState state = Block.BLOCK_STATE_IDS.getByValue(buf.readVarInt());
+                assert state != null;
+                states.put(state, i);
+            }
+            return states;
+        }
+    }
```

Comment:
because we need the complete pallete to do look ahead to see if we even need to scan the chunk... that doesn't work with getBlockState since it takes a location

Reply:
To me this looks like we can easily iterate over it
```java
    IBlockState getBlockState(int indexKey);
```
Only problem is that on 1.12.2 `IBlockStatePalette.size` doesn't exist so we'd either have to add it with a mixin or iterate until we get a null state (needs special handling of `BlcosStatePaletteRegsitry` because it returns air instead) or simply access the underlying storage directly. Copying should only be done if we don't want to assume out of bounds indices cause a null state and get an unknown palette implementation.

Reply:
really the copying is there because it's fast enough and writing a seperate implementation for each implementation of IBlockStatePalette is annoying... and there's more on newer mc versions.

> Copying should only be done if we don't want to assume out of bounds indices cause a null state and get an unknown palette implementation.

before this there was a lot more copying... https://github.com/cabaletta/baritone/blob/cfdbc851a005c38818cd204f51302bd920f7b1f4/src/launch/java/baritone/launch/mixins/MixinBitArray.java#L49


I don't really see the problem here. this part of the code isn't the hotest part and so it's good enough. it's not like removing this copying would lower the O notation. it's already O(n) on the code checking the pallete

Reply:
> before this there was a lot more copying...

Bad argument when implementing a speedup of multiple orders of magnitude but if you say it's fast enough with much less effort then I'll take that.